src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs

```diff
@@ -105,8 +154,142 @@ public WebRequestSession()
+                MaxAutomaticRedirections = MaximumRedirection;
+            }
+
+            // Check timeout setting (in seconds instead of milliseconds as in HttpWebRequest)
+            var timeSpanTimeout = timeoutSec is 0 ? TimeSpan.FromMilliseconds(Timeout.Infinite) : TimeSpan.FromSeconds(timeoutSec);
+            if (_client is not null && !timeSpanTimeout.Equals(_client.Timeout))
+            {
+                ResetClient();
+            }
+
+            newClient = _client is null;
+
+            if (newClient)
+            {
+                _client = CreateHttpClient(timeSpanTimeout);
+            }
+
+            return _client;
+        }
+
+        private HttpClient CreateHttpClient(TimeSpan timeout)
+        {
+            HttpClientHandler handler = new();
+
+            handler.CookieContainer = Cookies;
+            handler.AutomaticDecompression = DecompressionMethods.All;
+
+            // Set the credentials used by this request
+            if (UseDefaultCredentials)
+            {
+                // The UseDefaultCredentials flag overrides other supplied credentials
+                handler.UseDefaultCredentials = true;
+            }
+            else if (Credentials is not null)
+            {
+                handler.Credentials = Credentials;
+            }
+
+            if (_noProxy)
+            {
+                handler.UseProxy = false;
+            }
+            else if (Proxy is not null)
+            {
+                handler.Proxy = Proxy;
+            }
 
-            Certificates.Add(certificate);
+            if (Certificates is not null)
+            {
+                handler.ClientCertificates.AddRange(Certificates);
+            }
+
+            if (_skipCertificateCheck)
+            {
+                handler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
+                handler.ClientCertificateOptions = ClientCertificateOption.Manual;
+            }
+
+            handler.AllowAutoRedirect = _allowAutoRedirect;
+            if (_allowAutoRedirect && MaximumRedirection > 0)
+            {
+                handler.MaxAutomaticRedirections = _maxAutomaticRedirections;
+            }
+
+            handler.SslProtocols = (SslProtocols)_sslProtocol;
+
+            return new HttpClient(handler)
+            {
+                Timeout = timeout
+            };
+        }
+
+        private void SetClassVar<T>(ref T oldValue, T newValue) where T : class
+        {
+            if (oldValue != newValue)
+            {
+                ResetClient();
+                oldValue = newValue;
+            }
+        }
+
+        private void SetStructVar<T>(ref T oldValue, T newValue) where T : struct
+        {
+            if (!oldValue.Equals(newValue))
+            {
+                ResetClient();
+                oldValue = newValue;
+            }
+        }
+
+        private void ResetClient()
+        {
+            _client?.Dispose();
```

Comment:
Do we reset client on every changed parameter? If so it is not correct.

Reply:
We reset client on every parameter that would cause it to throw an exception if we tried to update it.

This includes anything that would impact the HttpClientHandler or the client itself (you can't change the timeout on the client once it has issued a request).

Note that there is a problem with the latest refactor that is breaking some tests. I am trying to track down what is wrong at the moment.

Reply:
Changes in https://github.com/PowerShell/PowerShell/pull/19190 resolve the test failures in my PR.